config/dependency.go

```diff
@@ -273,10 +282,26 @@ func getTerragruntOutput(dependencyConfig Dependency, terragruntOptions *options
 		return nil, true, errors.WithStackTrace(DependencyConfigNotFound{Path: targetConfig})
 	}
 
-	jsonBytes, err := runTerragruntOutputJson(terragruntOptions, targetConfig)
-	if err != nil {
-		return nil, true, err
+	// Acquire synchronization lock to ensure only one instance of output is called per config.
+	actualLock, _ := outputLocks.LoadOrStore(targetConfig, &sync.Mutex{})
+	defer actualLock.(*sync.Mutex).Unlock()
+	actualLock.(*sync.Mutex).Lock()
+
+	// Look up if we have already run terragrunt output for this target config
+	var jsonBytes []byte
+	rawJsonBytes, hasRun := jsonOutputCache.Load(targetConfig)
+	if hasRun {
+		util.Debugf(terragruntOptions.Logger, "%s was run before. Using cached output.", targetConfig)
+		jsonBytes = rawJsonBytes.([]byte)
+	} else {
+		newJsonBytes, err := runTerragruntOutputJson(terragruntOptions, targetConfig)
+		if err != nil {
+			return nil, true, err
+		}
+		jsonBytes = newJsonBytes
+		jsonOutputCache.Store(targetConfig, jsonBytes)
```

Comment:
Minor nit: consider extracting the locking and caching logic into a separate method that returns the JSON byte data. That'll make the code here easier to read and in the method simpler, as you'll be able to immediately `return rawJsonBytes`.

Reply:
Done: 8a97525